topdown/strings.go

```diff
@@ -227,6 +227,28 @@ func builtinReplace(a, b, c ast.Value) (ast.Value, error) {
 	return ast.String(strings.Replace(string(s), string(old), string(new), -1)), nil
 }
 
+func builtinReplacerReplace(a, b ast.Value) (ast.Value, error) {
+	oldnew, err := builtins.ArrayOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := builtins.StringOperand(b, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	on := make([]string, len(oldnew))
+	for i, s := range oldnew {
+		on[i] = string(s.Value.(ast.String))
+	}
+
+	r := strings.NewReplacer(on...)
```

Comment:
`strings.NewReplacer` will panic if the list contains an odd number of elements. Let's make the API use a map instead of an array so that this becomes a non-issue. I'd also prefer a map syntactically.

Reply:
Thank you for the review. Do you mean the input becomes a `map` and converts a `map` to an `array` inside the function and than pass it to `NewReplacer`?

One concern is that, IIUC map's don't have guaranteed orders so depending on the input, the output may become flaky.

Or am I misunderstanding the requested change?
Thank you!

Reply:
I was thinking something like:

```
strings.replace_n("the string input", {"th": "!!", "n": "&&"})
```

This would produce:

```
"!!e stri&&g i&&put"
```

> One concern is that, IIUC map's don't have guaranteed orders so depending on the input, the output may become flaky.

I see. I didn't realize that the replacer worked this way. I don't have too strong of an opinion here except that it's unclear what the built-in should do if there are an odd number of arguments. The options are:

1. Return an error -- this is not great because errors are unrecoverable
2. Be undefined
3. Drop the last argument

If we use a map then it's impossible for this to happen.

Reply:
Maybe I'm being too cautious here, but suppose there was a sting like this

```bash
c/r/u/d
```

and you want to replace it to 

```bash
create/read/update/delete
```

In this case, the replacing order matters. It will succeed in the following order:

1. d -> delete
2. r -> read
3. c -> create
4. u -> update

If you accidentally replace it starting from `c -> create`, the `r` gets replaced to `read` in the latter replacements (`create` becomes `creadeate`). This may lead to problems in the future.

>  it's unclear what the built-in should do if there are an odd number of arguments. The options are:

I agree. I'd go with option 3 `Drop the last argument`. This will need to be well documented but will be `the best it can do` solution. I don't have a strong background for this opinion though.

Option 1 sounds like the best style but not being able to recover seems too much of a disadvantage. Option 2 I think will lead to **hard to debug** undefined rules.

WDYT?

Reply:
I think we're ok on this. Looking at the source for the replacer it looks like it only iterates over the original string a single time and replaces substrings as it encounters them.

Reply:
@tsandall @patrick-east Terribly sorry about the confusion. I was somehow stereotyped with the `Replacements are performed in the order they appear in the target string` comment in the go docs. I've understood I was just misunderstanding it.

Should've experimented it myself before saying the wrong thing 😞 

I'll change the arguments to maps and thank you for the reviews.